Sherlock/replicator.h

```diff
@@ -0,0 +1,285 @@
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          internal_stop_(false),
+          remote_stream_(remote_stream,
+                         [this]() {
+                           TerminateSubscription();
+                           thread_.join();
+                         }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
+          TerminateSubscription();
+        }
+        thread_.join();
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    void Thread() {
+      ThreadImpl();
+      subscriber_thread_done_ = true;
+      if (done_callback_) {
+        done_callback_();
+      }
+    }
+
+    void ThreadImpl() {
+      const std::string url = remote_stream_.ObjectAccessorDespitePossiblyDestructing().url_;
+      bool terminate_sent = false;
+      while (!internal_stop_) {
+        if (!terminate_sent && external_stop_) {
+          terminate_sent = true;
+          if (subscriber_.Terminate() != ss::TerminationResponse::Wait) {
+            return;
+          }
+        }
+        try {
+          HTTP(ChunkedGET(
+              url + "?i=" + current::ToString(index_),
+              [this](const std::string& header, const std::string& value) { OnHeader(header, value); },
+              [this](const std::string& chunk_body) { OnChunk(chunk_body); },
+              [this]() {}));
+        } catch (current::Exception&) {
+        }
+      }
+    }
+
+    void OnHeader(const std::string& header, const std::string& value) {
+      if (header == "X-Current-Stream-Subscription-Id") {
+        std::unique_lock<std::mutex> lock(mutex_);
+        subscription_id_ = value;
+        event_.notify_one();
+      }
+    }
+
+    void OnChunk(const std::string& chunk) {
+      if (external_stop_ || internal_stop_) {
+        return;
+      }
+
+      const auto split = current::strings::Split(chunk, '\t');
+      assert(split.size() == 2u);
+      const auto idx = ParseJSON<idxts_t>(split[0]);
+      assert(idx.index == index_);
+      auto entry = ParseJSON<TYPE_SUBSCRIBED_TO>(split[1]);
+      internal_stop_ =
+          (subscriber_(std::move(entry), idx, idxts_t() /*unused "last" arg*/) == ss::EntryResponse::Done);
+      std::unique_lock<std::mutex> lock(mutex_);
+      ++index_;
+      event_.notify_one();
+    }
+
+    void TerminateSubscription() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      event_.wait(lock, [this]() { return !subscription_id_.empty(); });
```

Comment:
Then may be we should check `subscriber_thread_done_` here. But it also won't help if the thread would continuously fail to connect, for example.